### PR TITLE
Simplify chunk assembly

### PR DESCRIPTION
**Pull request by jpfr**

### CODE REVIEW

File: src/ua_securechannel.c [hunk 1]
```diff
@@ -690,83 +520,269 @@ decryptAddChunk(UA_SecureChannel *channel, UA_MessageType messageType,
...
+            break;
+        }
+
+        deleteChunks(&doneChunks); 
+        if(res != UA_STATUSCODE_GOOD)
+            break;
+
+        /* Could we process a full message? Done? */
+        UA_Chunk *next = SIMPLEQ_FIRST(&channel->completeChunks);
+        if(chunk == next)
```
Comment (mlgiraud) [inline src/ua_securechannel.c]: I think i understand what we are doing here, but clang analyzer thinks this condition is always true, which is not the case.
Might indicate a problem that i don't see or too complex code. (And maybe a bug in clang analyzer of course)

File: src/client/ua_client_connect.c [hunk 2]
```diff
@@ -394,27 +394,29 @@ decodeProcessOPNResponseAsync(void *application, UA_SecureChannel *channel,
...
         UA_Client_disconnect(client);
         return;
     }
 
+   /* Decode and verify the sequence header */
+    UA_SequenceHeader sequenceHeader;
+    retval = UA_SequenceHeader_decodeBinary(msg, &offset, &sequenceHeader);
 #ifndef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
-    retval = processSequenceNumberAsym(channel, sequenceNumber);
+    retval |= processSequenceNumberAsym(channel, sequenceHeader.sequenceNumber);
```
Comment (mlgiraud) [inline src/client/ua_client_connect.c]: OR'ing the status codes is a bad idea. If two bad status codes are OR'ed it is impossible to reconstruct both error codes in most cases.

File: src/ua_securechannel.c [hunk 3]
```diff
@@ -690,83 +520,269 @@ decryptAddChunk(UA_SecureChannel *channel, UA_MessageType messageType,
...
+                       void *application, UA_ProcessMessageCallback callback) {
+    /* Only MSG and CLO messages at this point */
+    UA_assert(messageType == UA_MESSAGETYPE_MSG ||
+              messageType == UA_MESSAGETYPE_CLO);
+
+    /* Assemble message */
+    UA_Chunk *chunk;
+    UA_ByteString payload;
+    if(chunksSize > 1) {
+        /* Compute the full message length. Omit the MessageHeader and
```
Comment (mlgiraud) [inline src/ua_securechannel.c]: We can avoid this loop by calculating the size when the chunks are extracted from the packet, or when the chunksSize variable is calculated.

**Comment by jpfr** [comment]
Thanks for taking a look.
Yes, let's bring this up after our next call.

File: src/ua_securechannel.c [hunk 4]
```diff
@@ -690,83 +520,269 @@ decryptAddChunk(UA_SecureChannel *channel, UA_MessageType messageType,
...
+        return UA_STATUSCODE_GOOD;
+
+    /* Compare that the messageType until the final chunk is the same */
+    UA_MessageType messageType = chunk->messageType;
+
+    /* Do we have a complete message? */
+    UA_ChunkType finalType = UA_CHUNKTYPE_INTERMEDIATE;
+    size_t chunksSize = 0;
+    UA_Chunk *finalChunk = NULL;
+    SIMPLEQ_FOREACH(finalChunk, &channel->completeChunks, pointers) {
```
[the remark this answers is not in the document]
Comment (jpfr) [reply src/ua_securechannel.c:654]: Integrate the message size count in here.

[the remark this answers is not in the document]
Comment (jpfr) [reply src/ua_securechannel.c:654]: But... padding is still integrated here.

**Pull request merged**